subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/GroovyRecompilationSpecProvider.java

```diff
@@ -96,49 +90,44 @@ public void initializeCompilation(JavaCompileSpec spec, RecompilationSpec recomp
-            filesToRecompilePatterns.include(relativize(file));
+    private void prepareFilePatterns(Set<String> relativeSourcePathsToCompile, PatternSet classesToDelete, PatternSet filesToRecompilePatterns) {
+        for (String fileRelativePath : relativeSourcePathsToCompile) {
+            filesToRecompilePatterns.include(fileRelativePath);
 
-            Collection<String> classes = sourceFileClassNameConverter.getClassNames(file);
-            for (String staleClass : classes) {
-                String path = staleClass.replaceAll("\\.", "/");
-                classesToDelete.include(path.concat(".class"));
-            }
+            sourceFileClassNameConverter.getClassNames(fileRelativePath)
+                .stream()
+                .map(staleClass -> staleClass.replaceAll("\\.", "/").concat(".class"))
+                .forEach(classesToDelete::include);
         }
     }
 
-    private String relativize(File sourceFile) {
-        List<File> dirs = sourceDirs.getSourceRoots();
-        for (File sourceDir : dirs) {
-            if (sourceFile.getAbsolutePath().startsWith(sourceDir.getAbsolutePath())) {
-                return RelativePathUtil.relativePath(sourceDir, sourceFile);
-            }
-        }
-        throw new IllegalStateException("Not found " + sourceFile + " in source dirs: " + dirs);
-    }
-
     private void processOtherChanges(PreviousCompilation previous, RecompilationSpec spec) {
         if (spec.getFullRebuildCause() != null) {
             return;
         }
-        SourceFileChangeProcessor sourceFileChangeProcessor = new SourceFileChangeProcessor(previous, sourceFileClassNameConverter);
+        SourceFileChangeProcessor sourceFileChangeProcessor = new SourceFileChangeProcessor(previous);
+
         for (FileChange fileChange : sourceChanges) {
             if (spec.getFullRebuildCause() != null) {
                 return;
             }
-            spec.getFilesToCompile().add(fileChange.getFile());
-            sourceFileChangeProcessor.processChange(fileChange.getFile(), spec);
+
+            File changedFile = fileChange.getFile();
+            String relativeFilePath = fileChange.getNormalizedPath();
+
+            Collection<String> changedClasses = sourceFileClassNameConverter.getClassNames(relativeFilePath);
+            spec.getRelativeSourcePathsToCompile().add(relativeFilePath);
+            sourceFileChangeProcessor.processChange(changedFile, changedClasses, spec);
         }
 
         for (String className : spec.getClassesToCompile()) {
             if (spec.getFullRebuildCause() != null) {
                 return;
             }
 
-            Optional<File> sourceFile = sourceFileClassNameConverter.getFile(className);
-            if (sourceFile.isPresent()) {
-                spec.getFilesToCompile().add(sourceFile.get());
+            Optional<String> relativeSourceFile = sourceFileClassNameConverter.getFileRelativePath(className);
+            if (relativeSourceFile.isPresent()) {
```

Comment:
You could use `relativeSourceFile.ifPresent`. Best would be `ifPresentOrElse`, though that is only available in Java 9+.
Using `ifPresent` avoids the call to get, though it is not much better since you still need to use `isPresent` for the `else` block.

Reply:
Yeah... so `ifPresent` doesn't help much here...